### PR TITLE
Add a check Callback to the Pattern Paritioner

### PR DESCRIPTION
**Pull request by mbrookhart**
cc @comaniac @mbaret @masahi

### CODE REVIEW

File: include/tvm/relay/dataflow_matcher.h [hunk 1]
```diff
@@ -90,7 +91,8 @@ Expr RewritePatterns(Array<DFPatternCallback> callbacks, Expr expr);
  *
  * \return Return the paritioned Expr.
  */
-Expr PartitionPattern(DFPattern pattern, Expr expr);
```
Comment (mbaret) [inline include/tvm/relay/dataflow_matcher.h:93]: The history's a bit confusing here, did the last version not have an attrs argument?

File: tests/python/relay/test_dataflow_pattern.py [hunk 2]
```diff
@@ -959,4 +984,5 @@ def test_parition_double_batchnorm():
     test_quadruple_partition_dominator()
     test_parition_batchnorm()
     test_parition_double_batchnorm()
+    test_parition_check()
```
Comment (mbaret) [inline tests/python/relay/test_dataflow_pattern.py:987]: test_partition

File: src/relay/ir/dataflow_matcher.cc [hunk 3]
```diff
@@ -728,16 +729,17 @@ class PatternPartitioner : protected MixedModeMutator {
...
+                      PackedFunc check) {
+  return PatternPartitioner().Partition(pattern, expr, attrs, check);
 }
 
 TVM_REGISTER_GLOBAL("relay.dataflow_pattern.partition")
-    .set_body_typed([](DFPattern pattern, Expr expr, Map<std::string, ObjectRef> attrs) {
-      return PartitionPattern(pattern, expr, attrs);
-    });
+    .set_body_typed([](DFPattern pattern, Expr expr, Map<std::string, ObjectRef> attrs,
+                       PackedFunc check) { return PartitionPattern(pattern, expr, attrs, check); });
```
Comment (mbaret) [inline src/relay/ir/dataflow_matcher.cc:742]: Probably keep the original formatting here? Not sure what the style convention is though but it definitely reads more clearly.

**Comment by comaniac** [comment]
> LGTM. I saw that you've fixed the debug print so I'll close #5642.

Ah TQ just merged that one so please rebase to the master lol

File: include/tvm/relay/dataflow_matcher.h (hunk 1 above)
Comment (mbrookhart) [reply include/tvm/relay/dataflow_matcher.h:93]: Yes, missed it in the last PR yesterday :(. Kind of surprised I didn't get a compiler warning.

File: src/relay/ir/dataflow_matcher.cc (hunk 3 above)
Comment (mbrookhart) [reply src/relay/ir/dataflow_matcher.cc:742]: Tianqi added clang-format-10 to the lint checker, this is the output of that version of clang-format.

File: tests/python/relay/test_dataflow_pattern.py (hunk 2 above)
Comment (mbrookhart) [reply tests/python/relay/test_dataflow_pattern.py:987]: Ooo, I should fix a few of those, thanks for noticing!

**Comment by mbrookhart** [comment]
@mbaret Let me take a look at that PR

File: src/relay/ir/dataflow_matcher.cc (hunk 3 above)
Comment (mbaret) [reply src/relay/ir/dataflow_matcher.cc:742]: Ah, fair enough. In clang-format we trust.

File: include/tvm/relay/dataflow_matcher.h (hunk 1 above)
Comment (mbaret) [reply include/tvm/relay/dataflow_matcher.h:93]: That is surprising, alright no worries then.

**Comment by mbrookhart** [comment]
@mbaret Okay, took a look at the type system. 

I'm actually passing you the original expression to check, not the partitioned version, so i don't think it will be a problem.

That being said, I am not propagating type information through partitioning, in part because I'm not sure at what stage of the compiler passes it will be used, and it would fail if someone tries to run this before type inference.

I am, however, I'm running InferTypes as part of the matcher for TypePatterns, so maybe I should just add the requirement that the inputs need to have well formed types and the pass will run InferTypes by default.

Thoughts?

**Comment by comaniac** [comment]
There is only one issue that we need to think of about the type propagation. The reason merge composite (and some other BYOC passes) needs to manually propagate type is that later processed nodes may need to check the type of its arguments (such as the check example in this PR). Since the argument node may be the CallNode that we created for the partitioned function, we need to set up the type to make the check function work.

On the other hand, I am thinking that if we make sure all argument accesses are based on the `pre` node instead of `post` node, then manual type propagation may not be an issue.

**Comment by mbaret** [comment]
We should have a test case to investigate this. Calling type infer after the partitioning pass probably deals with most of the issues that could arise.

**Comment by mbrookhart** [comment]
I'll write something based on the test case in your PR.

**Comment by mbrookhart** [comment]
@mbaret I'm noticing you don't use structural_equal in your tests after partitioning. I'm having issues with that pass when I introduce types. Does it not work for types?

**Comment by mbaret** [comment]
I'll be honest and say the reason I didn't use structural equal for the second check was because I didn't want to write out the partitioned Relay expression :) I figured for that particular test I just want to know whether or not the check worked, there are plenty of tests to check the validity of partitioning above.

I'm not entirely sure how structural_equal interacts with types, it's not a pass I've looked into in much detail.

**Comment by masahi** [comment]
@mbrookhart @mbaret is @mbaret's PR #5640 ready to merge?

**Comment by mbrookhart** [comment]
Yes! Using it as inspiration for future improvements, but there's nothing here that would block it.

**Comment by masahi** [comment]
Thanks @mbrookhart @mbaret

**Pull request merged**